src/natcap/invest/urban_flood_risk_mitigation.py

```diff
@@ -322,107 +313,245 @@ def execute(args):
+        task_name='write summary stats to flood_risk_service.shp',
+        dependent_task_list=summary_tasks)
 
     task_graph.close()
     task_graph.join()
 
 
-def _pickle_zonal_stats(
-        base_vector_path, base_raster_path, target_pickle_path):
-    """Calculate Zonal Stats for a vector/raster pair and pickle result.
+def _write_summary_vector(
+        source_aoi_vector_path, target_vector_path, runoff_ret_stats=None,
+        runoff_ret_vol_stats=None, damage_per_aoi_stats=None,
+        flood_volume_stats=None):
 
-    Parameters:
-        base_vector_path (str): path to vector file
-        base_raster_path (str): path to raster file to aggregate over.
-        target_pickle_path (str): path to desired target pickle file that will
-            be a pickle of the pygeoprocessing.zonal_stats function.
+    source_aoi_vector = gdal.OpenEx(source_aoi_vector_path, gdal.OF_VECTOR)
+    source_aoi_layer = source_aoi_vector.GetLayer()
+    source_geom_type = source_aoi_layer.GetGeomType()
+    source_srs_wkt = pygeoprocessing.get_vector_info(
+        source_aoi_vector_path)['projection']
+    source_srs = osr.SpatialReference()
+    source_srs.ImportFromWkt(source_srs_wkt)
 
-    Returns:
-        None.
+    esri_driver = gdal.GetDriverByName('ESRI Shapefile')
+    target_watershed_vector = esri_driver.Create(
+        target_vector_path, 0, 0, 0, gdal.GDT_Unknown)
+    layer_name = str(os.path.splitext(os.path.basename(
+        target_vector_path))[0])
+    LOGGER.debug("creating layer %s", layer_name)
+    target_watershed_layer = target_watershed_vector.CreateLayer(
+        str(layer_name), source_srs, source_geom_type)
+
+    target_fields = ['rnf_rt_idx', 'rnf_rt_m3']
+    if not damage_per_aoi_stats:
+        damage_per_aoi_stats = {}
+    else:
+        target_fields += ['aff_bld', 'serv_bld']
+
+    for field_name in target_fields:
+        field_def = ogr.FieldDefn(field_name, ogr.OFTReal)
+        field_def.SetWidth(36)
+        field_def.SetPrecision(11)
+        target_watershed_layer.CreateField(field_def)
+
+    target_layer_defn = target_watershed_layer.GetLayerDefn()
+    for base_feature in source_aoi_layer:
+        feature_id = base_feature.GetFID()
+        target_feature = ogr.Feature(target_layer_defn)
+        base_geom_ref = base_feature.GetGeometryRef()
+        target_feature.SetGeometry(base_geom_ref.Clone())
+        base_geom_ref = None
+
+        if feature_id in runoff_ret_stats:
+            pixel_count = runoff_ret_stats[feature_id]['count']
+            if pixel_count > 0:
+                mean_value = (
+                    runoff_ret_stats[feature_id]['sum'] / float(pixel_count))
+                target_feature.SetField('rnf_rt_idx', float(mean_value))
+
+        if feature_id in runoff_ret_vol_stats:
+            target_feature.SetField(
+                'rnf_rt_m3', float(
+                    runoff_ret_vol_stats[feature_id]['sum']))
+
+        if feature_id in damage_per_aoi_stats:

```

Comment:
> is it important to check that feature_id is in all of them?
I think we can safely assume that under normal execution of the model the feature ID will be present in the stats dict.  This is because this task and the `zonal_statistics` task that created the stats dict are both run with the same watersheds/aoi vector.  There could, of course, be a case that I'm not expecting here, but @davemfish what do you think about waiting to see what that case looks like before guarding against it?

Reply:
> Also I'm a little confused why the pixel count is taken from flood_volume_stats while the other stats are from other datasets.

Ha, good catch!  I've patched this in 14f02eb9 so that the pixel count is taken from another stats dict.  This also has the convenient side effect of doing one less `zonal_statistics`. 

> Is this function missing a block for writing flood_volume_stats?

Based on the UG, I can only guess at why the `flood_volume_stats` were being used in a previous version.  I'll confirm with Perrine that we should not be outputting some summary statistic of the flood volume.

And for future us, just back out of 14f02eb9 to restore the removed bits!

Reply:
Yeah, seems like a good idea to double-check. I don't know this model deeply at all, but now I wonder why the flood volume raster is calculated at all if it's not meant to be summarized and it doesn't appear to be an intermediate step towards something else. Maybe it's sometimes useful to the modeler though.